### PR TITLE
[#133] feat(netty): Introduce ShuffleServerGrpcNettyClient.

### PR DESCRIPTION
**Pull request by leixm**
### What changes were proposed in this pull request?
Introduce ShuffleServerGrpcNettyClient.

### Why are the changes needed?

for #133 

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

existing UTs.

### CODE REVIEW

**Comment by leixm** [comment]
cc @smallzhongfeng  PTAL.

**Comment by leixm** [comment]
It seems a bit messy at the moment. Some codes use ByteBuffer, such as FixedSizeSegmentSplitter and ShuffleReadClientImpl, and some codes use ByteBuf, such as netty client or protocol objects. At present, I prefer to use ByteBuf in nettyclient, and keep ByteBuffer in other places.

**Comment by smallzhongfeng** [comment]
> It seems a bit messy at the moment. Some codes use ByteBuffer, such as FixedSizeSegmentSplitter and ShuffleReadClientImpl, and some codes use ByteBuf, such as netty client or protocol objects. At present, I prefer to use ByteBuf in nettyclient, and keep ByteBuffer in other places.

+1 for this change. but also cc @jerqi

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcNettyClient.java [hunk 1]
```diff
@@ -0,0 +1,253 @@
...
+              requestId(),
+              request.getAppId(),
+              shuffleId,
+              requireId,
+              stb.getValue(),
+              System.currentTimeMillis());
+          long start = System.currentTimeMillis();
+          RpcResponse rpcResponse = transportClient.sendRpcSync(sendShuffleDataRequest, RPC_TIMEOUT_DEFAULT_MS);
+          LOG.debug("Do sendShuffleData to {}:{} rpc cost:" + (System.currentTimeMillis() - start)
+                        + " ms for " + allocateSize + " bytes with " + finalBlockNum + " blocks", host, port);
```
Comment (smallzhongfeng) [inline internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcNettyClient.java:104]: nit: code Indent

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcNettyClient.java [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+          if (rpcResponse.getStatusCode() != StatusCode.SUCCESS) {
+            String msg = "Can't send shuffle data with " + finalBlockNum
+                             + " blocks to " + host + ":" + port
```
Comment (smallzhongfeng) [inline internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcNettyClient.java:107]: nit: code Indent

File: internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcNettyClient.java [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+        request.getAppId(),
+        request.getShuffleId(),
+        request.getPartitionId(),
+        request.getLastBlockId(),
+        request.getReadBufferSize(),
+        System.currentTimeMillis(),
+        request.getExpectedTaskIds()
+    );
+    String requestInfo = "appId[" + request.getAppId()
+                             + "], shuffleId[" + request.getShuffleId()
```
Comment (smallzhongfeng) [inline internal-client/src/main/java/org/apache/uniffle/client/impl/grpc/ShuffleServerGrpcNettyClient.java:149]: ditto

**Review by smallzhongfeng — COMMENTED** [review]
LGTM except for some minor changes.

**Comment by jerqi** [comment]
> It seems a bit messy at the moment. Some codes use ByteBuffer, such as FixedSizeSegmentSplitter and ShuffleReadClientImpl, and some codes use ByteBuf, such as netty client or protocol objects. At present, I prefer to use ByteBuf in nettyclient, and keep ByteBuffer in other places.

OK for me.

**Comment by leixm** [comment]
@smallzhongfeng  @jerqi  PTAL.

**Review by jerqi — APPROVED** [review]
LGTM

**Review by smallzhongfeng — APPROVED** [review]
LGTM, thanks @leixm

**Pull request merged**